File: src/android/com/eclipsesource/tabris/maps/MapCameraChangeListener.java
package com.eclipsesource.tabris.maps;

import com.eclipsesource.tabris.client.core.ObjectRegistry;
import com.google.android.gms.maps.GoogleMap.OnCameraIdleListener;
import com.google.android.gms.maps.GoogleMap.OnCameraMoveStartedListener;
import com.google.android.gms.maps.model.CameraPosition;

import java.util.HashMap;
import java.util.Map;

import static java.util.Arrays.asList;

public class MapCameraChangeListener implements OnCameraMoveStartedListener, OnCameraIdleListener {

  public static final String EVENT_CAMERA_MOVE = "cameramove";
  public static final String EVENT_CHANGE_CAMERA = "changecamera";

  private final ObjectRegistry objectRegistry;
  private final MapHolderView mapHolderView;
  private int reason;

  public MapCameraChangeListener( MapHolderView mapHolderView, ObjectRegistry objectRegistry ) {
    this.objectRegistry = objectRegistry;
    this.mapHolderView = mapHolderView;
  }

  @Override
  public void onCameraMoveStarted( int reason ) {
    this.reason = reason;
  }

  @Override
  public void onCameraIdle() {
    CameraPosition cameraPosition = mapHolderView.getGoogleMap().getCameraPosition();
    notifyChangeCameraEvent( cameraPosition );
    if( reason == OnCameraMoveStartedListener.REASON_GESTURE ) {
      notifyCameraMoveEvent( cameraPosition );
    }
  }

  private void notifyChangeCameraEvent( CameraPosition cameraPosition ) {
    Map<String, Object> arguments = new HashMap<>();
    arguments.put( "position", asList( cameraPosition.target.latitude, cameraPosition.target.longitude ) );
    objectRegistry.getRemoteObjectForObject( mapHolderView ).notify( EVENT_CHANGE_CAMERA, arguments );
  }

  private void notifyCameraMoveEvent( CameraPosition cameraPosition ) {
    Map<String, Object> arguments = new HashMap<>();
    arguments.put( "position", asList( cameraPosition.target.latitude, cameraPosition.target.longitude ) );
    objectRegistry.getRemoteObjectForObject( mapHolderView ).notify( EVENT_CAMERA_MOVE, "camera", arguments );
  }

}

File: src/android/com/eclipsesource/tabris/maps/MapCameraChangeListener.kt
package com.eclipsesource.tabris.maps

import com.eclipsesource.tabris.android.Scope
import com.google.android.gms.maps.GoogleMap.OnCameraIdleListener
import com.google.android.gms.maps.GoogleMap.OnCameraMoveStartedListener
import com.google.android.gms.maps.GoogleMap.OnCameraMoveStartedListener.REASON_API_ANIMATION
import com.google.android.gms.maps.GoogleMap.OnCameraMoveStartedListener.REASON_GESTURE
import com.google.android.gms.maps.model.CameraPosition

class MapCameraChangeListener(private val mapHolderView: MapHolderView, private val scope: Scope)
  : OnCameraMoveStartedListener, OnCameraIdleListener {

  private var reason: Int = 0

  override fun onCameraMoveStarted(reason: Int) {
    this.reason = reason
  }

  override fun onCameraIdle() {
    mapHolderView.googleMap.cameraPosition?.let {
      notifyChangeCameraEvent(it)
      when (reason) {
        REASON_GESTURE, REASON_API_ANIMATION -> notifyCameraMoveEvent(it)
      }
    }
  }

  private fun notifyChangeCameraEvent(cameraPosition: CameraPosition) {
    scope.remoteObject(mapHolderView)?.notify("changeCamera", mapOf(
        "position" to listOf(cameraPosition.target.latitude, cameraPosition.target.longitude)
    ))
  }

  private fun notifyCameraMoveEvent(cameraPosition: CameraPosition) {
    scope.remoteObject(mapHolderView)?.notify("cameraMoved", "camera", mapOf(
        "position" to listOf(cameraPosition.target.latitude, cameraPosition.target.longitude)
    ))
  }

}
